[FEAT] 채팅방 생성 로직 구현

dye0p opened the pull request:
## Issue Number
closed #811

## As-Is
<!-- 문제 상황 정의 -->
멘토링 예약이 승인되어도 멘티와 멘토가 소통할 수 있는 채팅방이 생성되지 않았습니다. 따라서, 멘토링 진행을 위해 별도의 외부 소통 채널을 이용해야 하는 불편함이 있었습니다.

## To-Be
<!-- 변경 사항 -->
 멘토링 예약이 승인되면 해당 멘토링을 위한 채팅방이 자동으로 생성됩니다. 멘티는 예약 승인 알림과 함께 생성된 채팅방 URL을 SMS로 수신하게 됩니다, 이제 멘토링을 위한 채팅방에 접속할 수 있습니다.

- Reservation 상태가 APPROVE로 변경될 때 ChatRoom을 생성하고, 생성된 채팅방의 URL을 반환하도록 로직을 추가했습니다.
- ChatRoom 엔티티와 ChatRoomRepository를 추가했습니다.
- 생성된 채팅방 ID를 기반으로 고유한 URL을 생성하는 ChatRoomUrlGenerator를 구현했습니다.
- 예약 승인 시, SMS 알림 메시지에 채팅방 접속 URL을 포함하도록 SmsMessageFormatter를 수정했습니다.
- chat_room 테이블 생성을 위한 DB 마이그레이션 스크립트를 추가했습니다.
- chatRoom 엔티티의 연관관계(Reservation, Member(mentor, mentee)는 객체 직접 참조 대신 간접 참조하도록 구현하였습니다.

### chat_room 테이블 추가
<img width="390" height="137" alt="image" src="https://github.com/user-attachments/assets/a8e591f9-3ff2-42bf-b421-cd3633abb403" />

JPA 상에서는 연관관계를 맺지 않지만, DB상에서는 연관관계가 있다고 판단하여 연관관계 매핑을 해주었습니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## (Optional) Additional Description

### 개선 필요 사항
현재 ReservationService의 updateStatus 메서드는 다음과 같은 구조를 가지고 있습니다
<img width="1976" height="935" alt="image" src="https://github.com/user-attachments/assets/adb275b4-34d6-41ce-be52-eb92e6cdf034" />


이 방식에는 다음과 같은 문제가 존재합니다:
	1.	분기 처리 증가
	•	상태(Status)에 따라 분기 처리가 필요하며, 기능이 확장될수록 if 문이 계속 늘어납니다.
	•	예: 추후 “거절” 상태에서 추가 로직이 생기면 또 다른 분기문 추가 필요.
	2.	메서드 이름과 행동의 불일치
	•	updateStatus라는 이름은 단순히 상태를 변경하는 역할처럼 보이지만, 내부에서 채팅방 생성 등 부수 효과가 발생합니다.
	3.	계약(Contract)이 모호함
	•	메서드가 반환하는 값이 상태에 따라 달라집니다.
	•	어떤 상태일 때 어떤 반환값을 기대할 수 있는지 외부에서 알기 어렵습니다.
	4.	확장성과 테스트성 저하
	•	새로운 상태가 추가될 때마다 로직이 비대해지고, 단위 테스트가 어려워집니다.

### 개선 제안

상태별 책임을 명확히 하기 위해 메서드를 분리합니다.
approve, reject 등 구체적인 행위를 나타내는 메서드로 나누어 명시적이고 예측 가능한 행동을 보장합니다.
예시 
``` java
/**
 * 예약을 승인하고, 생성된 채팅방 정보를 반환합니다.
 */
@Transactional
public ApprovedReservationInfo approve(Long reservationId) {
    Reservation reservation = getReservation(reservationId);
    reservation.approve(); // 도메인 객체가 스스로 상태를 변경

    ChatRoomCreatedInfo chatRoomInfo = chatRoomService.registerChatRoom(reservation);

    // ‘승인’은 항상 채팅방 URL을 포함한 정보를 반환한다는 계약을 명확히 함
    return new ApprovedReservationInfo(reservation.getId(), chatRoomInfo.url());
}

/**
 * 예약을 거절합니다.
 */
@Transactional
public void reject(Long reservationId, String reason) {
    Reservation reservation = getReservation(reservationId);
    reservation.reject(reason); // 거절 사유와 함께 상태 변경
    // 거절 시 별도의 반환값이 필요하지 않음
}
```

필요 작업
- 각각의 예약 상태를 처리하는 메서드 구현
- 각각의 엔드포인트도 필요함
- 프론트 엔드의 요청 방식도 변경될 필요가 있어 논의 필요

dye0p committed:
feat: 채팅방(ChatRoom) 엔티티 및 스키마 추가

- chat_room 테이블 생성 (isDeleted, deletedAt 포함)
- ChatRoom 엔티티 클래스 추가
- @SQLDelete 적용으로 delete() 호출 시 soft delete 구현
- @SQLRestriction 적용으로 삭제되지 않은 레코드만 조회

dye0p committed:
feat: 채팅방 URL 생성 유틸 추가

- 채팅방 ID 기반 URL 생성 메서드 추가

dye0p committed:
feat: 채팅방 생성 service 로직 구현

- 채팅방 생성 시 채팅방 URL을 생성하여 DTO로 반환.

dye0p committed:
feat: 예약 승인 시 채팅방 생성 로직 추가

- 변경된 예약의 상태가 승인이라면 채팅방을 생성하여 예약과 채팅방 URL을 DTO에 담아 반환.

dye0p committed:
feat: 멘티에게 전송될 sms 메세지 인자에 채팅방 URL 추가

dye0p committed:
test: 예약 상태 업데이트에 따라서 채팅방 url 생성 테스트 추가

dye0p commented:
@CodeRabbitai 리뷰해

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/ChatRoomUrlGenerator.java:
```diff
@@ -0,0 +1,10 @@
+
+public class ChatRoomUrlGenerator {
+
+    private final static String BASE_URL = "https://www.fittoring.com/chat/room/";
```
p2: `private final static` 대신 `private static final`으로 통일하면 좋을 것 같아요~

ppparkta replied:
p2: 서브 도메인 부분을 환경변수로 관리해야 할 것 같습니다. (dev/prod)

```
private final static String BASE_URL = "https://" + ${서브도메인} +".fittoring.com/chat/room/";
```

dye0p replied:
> private final static 대신 private static final으로 통일하면 좋을 것 같아요~

그렇네요! 수정했습니다~

dye0p replied:
> p2: 서브 도메인 부분을 환경변수로 관리해야 할 것 같습니다. (dev/prod)
> 
> ```
> private final static String BASE_URL = "https://" + ${서브도메인} +".fittoring.com/chat/room/";
> ```

그렇네요! 알려주셔서 감사합니다. 수정했습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/ChatRoomUrlGenerator.java:
```diff
@@ -0,0 +1,10 @@
+    private final static String BASE_URL = "https://www.fittoring.com/chat/room/";
+
+    public static String generate(Long chatRoomId) {
+        return BASE_URL + chatRoomId;
```
p3: 이 url로 입장하면 웹소켓 연결이 시작되는 걸로 예상하면 될까요??

dye0p replied:
맞아요 해당 url은 사용자가 생성된 채팅방에 대해서 입장하는 url입니다. url로 접속하게 되면 클라이언트가 웹소켓 연결을 시도하게 됩니다~

kimdevv reviewed (approved):
채팅과 웹 소켓 관련해서 공부해야 할 부분이 많을 것 같은데 고생하셨습니다.
간단한 커멘트 남겨 놨고, 승인하겠습니다~

> 상태별 책임을 명확히 하기 위해 메서드를 분리합니다.
approve, reject 등 구체적인 행위를 나타내는 메서드로 나누어 명시적이고 예측 가능한 행동을 보장합니다.

그리고 이 부분은 저는 찬성합니다~ 상태가 많아질수록 분기문이 비대해질 것 같네요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/domain/model/ChatRoom.java:
```diff
@@ -0,0 +1,58 @@
+    @Id
+    private Long id;
+
+    @Column(name = "reservation_id", nullable = false)
+    private Long reservationId;
```
### p1: 채팅방이 예약과 연관관계를 맺는 것에 대해 논의가 필요합니다.

1차 스프린트 회의에서 멘토와 멘티가 여러 번 예약을 생성할 수 있고, 동일한 인물들 간의 대화라면 이전 대화기록을 볼 수 있게 하자고 논의했습니다.

예약과 연관관계를 맺고, 예약이 완료되는 시점에 채팅방이 생성되는 구조라면 이전의 대화 기록을 유지하기 힘들 것 같습니다. 

이에 따라 채팅방 구조를 함께 논의해보면 좋을 것 같습니다.

ppparkta replied:
p3: 연관관계 직접 맺지 않고 id만 갖고 있는 구조 좋네요 👍

jye326 replied:
p2 : 동의합니다. 아예 ChatRoom에 Reservation을 제거하고 mentee와 mentor의 id 쌍으로 채팅방 Url을 관리하는 것은 어떨까요?

dye0p replied:
> 1차 스프린트 회의에서 멘토와 멘티가 여러 번 예약을 생성할 수 있고, 동일한 인물들 간의 대화라면 이전 대화기록을 볼 수 있게 하자고 논의했습니다.

이 요구사항에 대해서 생각해 봤을 때 장단점이 존재할 것 같습니다.

**장점**
- 멘토와의 대화기록을 하나의 채팅방에서 관리할 수 있다(이전 대화 기록 찾기 편함)
- 같은 멘토에 대해서 채팅방이 여러개로 분산되지 않는다.

**단점**
-  대화 내용이 여러 멘토링 세션에 걸쳐 쌓이게 된다. 멘티는 특정 메시지가 어떤 멘토링 예약과 관련된 것인지 구분하기 어려울 수 있다.
- 멘토가 서로다른 성격의 멘토링을 개설했을 때, 특정 메세지가 어떤 멘토링과 관련된 것인지 구분하기 어려울 수 있다.

**궁금한 부분**
멘토링 프로그램 기준 (멘토링 타입 기반)으로 채팅방을 생성하는 것인지?
- a트레이너:  1멘토링, 2멘토링 개설
- d멘티: 1,2 멘토링 모두 예약  
   - 하나의 채팅방으로 관리?
   - 멘토링 단위로 채팅방 관리?

어떻게 하는 것이 좋을지 의견 남겨주세요~

ppparkta replied:
채팅 자체는 예약을 기준으로 분리하지만 채팅방은 멘토와 멘티 쌍으로 분리할 수도 있을 것 같아요! 

채팅방 -> 멘토, 멘티, List<채팅>
채팅 -> 발신자, 메시지, 예약

- 장점
  - 하나의 채팅방에서 여러 예약에 대한 채팅을 보여줄 수 있다
  - 특정 메시지가 어떤 예약에 관련된 것인지 분류하기 쉽다
- 단점
  - 조회 시 복잡하다

이렇게 정리해볼 수 있을 것 같아요! 단점이 크다면 위에서 띠용이 남겨준대로 지금은 멘토-멘티 쌍으로 묶는 것이 좋을 것 같습니다.

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/ChatRoomService.java:
```diff
@@ -0,0 +1,28 @@
+        ChatRoom chatRoom = new ChatRoom(
+                reservation.getId(),
+                reservation.getMentee().getId(),
+                reservation.getMentoring().getMentor().getId()
```
p3: 디미터 법칙 적용하면 좋을 것 같습니다

dye0p replied:
네 변경하였습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/application/reservation/service/MentoringReservationFacadeService.java:
```diff
@@ -20,7 +21,11 @@ public ReservationCreateResponse reserveMentoring(ReservationCreateDto dto) {
+        reservationNotificationService.sendReservationStatusUpdateSmsMessage(
+                reservationInfo.reservation(),
+                reservationInfo.reservation().getStatus(),
+                reservationInfo.chatRoomUrl()
```
p3: 👍

ppparkta reviewed (commented):
고생 많으셨습니다. 채팅 작업 단위가 굉장히 클텐데 하나씩 잘 해주고 계셔서 든든하네요.
채팅방 테이블 구조는 당장 떠오르는 것이 없어서 아이디어를 내지 못했는데 함께 이야기해보면 좋을 것 같아요!

jye326 commented on backend/fittoring/src/main/java/fittoring/application/chatroom/service/ChatRoomService.java:
```diff
@@ -0,0 +1,28 @@
+    private final ChatRoomRepository chatRoomRepository;
+
+    @Transactional
+    public ChatRoomCreatedInfo registerChatRoom(Reservation reservation) {
```
p3 : Reservation 대신 Mentor와 Mentee의 아이디로 채팅방을 등록하는 것은 어떨까요?

dye0p replied:
그렇게 제안하신 이유가 궁금합니다!

jye326 replied:
언급해주신 장점 때문입니다. 말씀해주신 것처럼 멘토 멘티의 아이디를 포함하는 개념인 `Metoring` 단위로 채팅방을 관리하는 것이 좋을 것 같아요!

> 장점
> 
> 멘토와의 대화기록을 하나의 채팅방에서 관리할 수 있다(이전 대화 기록 찾기 편함)
> 같은 멘토에 대해서 채팅방이 여러개로 분산되지 않는다.

> 단점
> 
> 대화 내용이 여러 멘토링 세션에 걸쳐 쌓이게 된다. 멘티는 특정 메시지가 어떤 멘토링 예약과 관련된 것인지 구분하기 어려울 수 있다.
> 멘토가 서로다른 성격의 멘토링을 개설했을 때, 특정 메세지가 어떤 멘토링과 관련된 것인지 구분하기 어려울 수 있다.

### 단점에 대해서 생각해 본 바로는
먼저 멘토링 예약마다 제공하는 서비스가 다른 경우가 아닙니다. 같은 서비스의 연장이죠. 시간이 다 되면 코인을 더 넣는 코인노래방에 비유할 수 있겠네요. 더 부르려고 코인을 더 넣었는데 그때마다 방을 옮겨야 하면 불편할 것 같습니다. 또한 특정 메시지가 어떤 예약에 포함되었는지 구별할 필요도 없다고 생각합니다. 코인 노래방의 예시로 '내가 이 노래를 어느 방에서 불렀지?'는 사용자에게 중요하지 않을 것 같아요!

두 번째로 우려하신 단점은 멘토가 여러 개의 멘토링을 개설할 수 있다는 확장성에 기반한 고민인 것 같습니다.  이 점도 현재 정책 상으로 크게 고려될 부분은 아니라고 생각합니다. 나중에 정책이 변경되고 수정되어도 괜찮을 것 같아요!

jye326 reviewed (commented):
고생하셨습니다! 채팅방의 연관 관계에 대한 논의가 한번 있으면 좋을 것 같아요! 코멘트 확인과 develop 충돌 해결 부탁드려요!

dye0p committed:
refactor: 상수 선언 방식 수정

- BASE_URL의 선언 순서를 Java 컨벤션에 맞게 변경 (static final 순서)

dye0p committed:
feat: 환경 변수 기반 채팅방 URL 동적 생성 적용

- ChatRoomUrlGenerator에 @Value를 이용해 sub-domain 설정값 주입
- 기존 정적 url 값을 동적으로 변경
- application.yml에 domain.sub-domain 설정값 추가
- ChatRoomService에서 채팅방 URL 생성 로직 수정

dye0p committed:
test: ReservationServiceTest에 채팅방 URL 관련 테스트 수정

- ChatRoomUrlGenerator를 ReservationServiceTest에 추가
- chatRoomUrl() 테스트에서 URL의 prefix 검증 수정 (`https://www` → `https://`)
- application-test.yml에 domain.sub-domain 설정값 추가

dye0p committed:
refactor: Reservation에서 멘토 조회 메서드 추가 및 사용 변경

- Reservation에 getMentor() 메서드 추가
- ChatRoomService에서 기존 reservation.getMentoring().getMentor() 호출을 reservation.getMentor()로 변경

dye0p reviewed (commented):
리뷰 반영과 의견 몇가지를 코멘트로 남겼습니다! 확인 해주세요~

dye0p committed:
Merge branch 'develop' into feature/811-create-chatroom

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/application/mentoring/repository/ChatRoomRepository.java
#	backend/fittoring/src/main/java/fittoring/application/mentoring/service/ChatRoomService.java
#	backend/fittoring/src/main/java/fittoring/application/mentoring/service/ChatRoomUrlGenerator.java
#	backend/fittoring/src/main/java/fittoring/application/mentoring/service/dto/ReservationInfo.java
#	backend/fittoring/src/main/java/fittoring/application/mentoring/service/dto/chat/ChatRoomCreatedInfo.java
#	backend/fittoring/src/main/java/fittoring/application/reservation/service/MentoringReservationFacadeService.java
#	backend/fittoring/src/main/java/fittoring/application/reservation/service/ReservationService.java
#	backend/fittoring/src/main/java/fittoring/domain/model/ChatRoom.java
#	backend/fittoring/src/main/resources/application.yml
#	backend/fittoring/src/test/java/fittoring/mentoring/business/service/ReservationServiceTest.java
#	backend/fittoring/src/test/resources/application-test.yml

ppparkta commented on backend/fittoring/src/main/java/fittoring/application/mentoring/service/ChatRoomUrlGenerator.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class ChatRoomUrlGenerator {
+
+    private static final String PROTOCOL = "https://";
+    private static final String DOMAIN = ".fittoring.com";
+    private static final String CHAT_PATH = "/chat/room/";
+
+    @Value("${domain.sub-domain}")
+    private String subDomain;
+
+
+    public String generate(Long chatRoomId) {
+        return PROTOCOL + subDomain + DOMAIN + CHAT_PATH + chatRoomId;
+    }
+}
```
오! 좋네요ㅎㅎ 사실 도기 코드를 다시 보면서 정작 제 코드에 환경변수 분리를 놓치고 있었다는 것을 알게됐어요. 덕분에 배워갑니다! 바꿔주신 방식 좋네요 👍

ppparkta reviewed (approved):
수정된 코드 확인했는데 좋네요! 채팅방 구조는 논의가 더 필요해보이는데 여태까지 나온 내용 토대로 의견 내봤습니다. 고생하셨어요!

dye0p committed:
Merge branch 'develop' into feature/811-create-chatroom

# Conflicts:
#	backend/fittoring/src/main/resources/application.yml
#	backend/fittoring/src/test/resources/application-test.yml
